src/fft_fftw3.f90

```diff
@@ -253,13 +276,17 @@ subroutine c2r_1m_z_plan(plan1, decomp_sp, decomp_ph)
       return
    end subroutine c2r_1m_z_plan
 
-!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    !  This routine performs one-time initialisations for the FFT engine
-!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-   subroutine init_fft_engine
+   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   subroutine init_fft_engine(engine)
 
       implicit none
 
+      type(decomp_2d_fft_engine), target, intent(inout) :: engine
```

Comment:
See above wrt a base type that is extended by different backends. We might then need each backend to have a "generic" subroutine that calls its specific one
```
subroutine init_fft_engine(engine)
  implicit none
  class(decomp_2d_fft_engine), target, intent(inout) :: engine ! This is generic, can be called by anyone

  call subroutine init_fftw3_engine(engine)
end subroutine


subroutine init_fftw3_engine(engine)
  implicit none
  type(decomp_2d_fftw3_engine), target, intent(inout) :: engine ! This is specific to the fftw3 type
  ...
end subroutine

Reply:
I am not sure about the exact syntax for extended types. It is not widely used in the code currently. This conversation is similar to https://github.com/2decomp-fft/2decomp-fft/pull/302#discussion_r1415630725. I will resolve this one and leave the one above open.